File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-shader-require-matching-varyings.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import {
  collectGlslGlobalDeclarations,
  type GlslGlobalDeclaration,
} from "./utils/collect-glsl-global-declarations.js";
import { hasGlslArrayDimensionMismatch } from "./utils/has-glsl-array-dimension-mismatch.js";
import { resolveStaticThreeShaderMaterial } from "./utils/resolve-static-three-shader-material.js";

const isVertexOutput = (declaration: GlslGlobalDeclaration): boolean =>
  declaration.qualifiers.has("out") || declaration.qualifiers.has("varying");

const isFragmentInput = (declaration: GlslGlobalDeclaration): boolean =>
  declaration.qualifiers.has("in") || declaration.qualifiers.has("varying");

const getInterfaceMismatch = (
  vertexOutput: GlslGlobalDeclaration,
  fragmentInput: GlslGlobalDeclaration,
): string | null => {
  if (vertexOutput.typeName !== fragmentInput.typeName) {
    return `type ${vertexOutput.typeName} in the vertex shader but ${fragmentInput.typeName} in the fragment shader`;
  }
  if (hasGlslArrayDimensionMismatch(vertexOutput.arraySize, fragmentInput.arraySize)) {
    return "different array dimensions";
  }
  return vertexOutput.interpolation === fragmentInput.interpolation
    ? null
    : `interpolation ${vertexOutput.interpolation} in the vertex shader but ${fragmentInput.interpolation} in the fragment shader`;
};

export const threeShaderRequireMatchingVaryings = defineRule({
  id: "three-shader-require-matching-varyings",
  title: "Shader stage interface does not match",
  category: "Correctness",
  severity: "error",
  recommendation:
    "Declare every statically used fragment input as a compatible vertex output with the same name",
  create: (context: RuleContext) => ({
    NewExpression(node: EsTreeNodeOfType<"NewExpression">) {
      const material = resolveStaticThreeShaderMaterial(node, context);
      if (!material?.vertexShader || !material.fragmentShader) return;
      const vertexOutputs = new Map(
        collectGlslGlobalDeclarations(material.vertexShader.program)
          .filter(isVertexOutput)
          .map((declaration) => [declaration.name, declaration]),
      );
      const fragmentInputs = collectGlslGlobalDeclarations(material.fragmentShader.program).filter(
        (declaration) =>
          isFragmentInput(declaration) &&
          declaration.isStaticallyUsed &&
          !declaration.name.startsWith("gl_") &&
          !declaration.hasLayoutQualifier,
      );
      for (const fragmentInput of fragmentInputs) {
        const vertexOutput = vertexOutputs.get(fragmentInput.name);
        const mismatch = vertexOutput
          ? getInterfaceMismatch(vertexOutput, fragmentInput)
          : "no matching vertex output";
        if (!mismatch) continue;
        context.report({
          node: material.fragmentShader.source.getOriginNodeAtOffset(
            fragmentInput.node.location?.start.offset ?? 0,
          ),
          message: `Fragment input ${fragmentInput.name} has ${mismatch}, so the shader program cannot link with a defined value`,
        });
      }
    },
  }),
});

File: native/oxlint/rules/three-shader-require-matching-varyings.rs
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use rustc_hash::FxHashMap;

use super::three_shader_require_matching_uniforms::{
    GlslGlobalDeclaration, for_each_static_matching_shader_material, has_array_dimension_mismatch,
};
use crate::{context::LintContext, rule::Rule};

#[derive(Debug, Default, Clone)]
pub struct ThreeShaderRequireMatchingVaryings;

declare_oxc_lint!(
    /// Require compatible vertex outputs and fragment inputs.
    ThreeShaderRequireMatchingVaryings,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Shader stage interface does not match.",
);

impl Rule for ThreeShaderRequireMatchingVaryings {
    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        for_each_static_matching_shader_material(
            ctx,
            |_vertex_shader, fragment_shader, vertex_declarations, fragment_declarations, ctx| {
                let vertex_outputs = vertex_declarations
                    .iter()
                    .filter(|declaration| is_vertex_output(declaration))
                    .map(|declaration| (declaration.name.as_str(), declaration))
                    .collect::<FxHashMap<_, _>>();
                let fragment_inputs = fragment_declarations.iter().filter(|declaration| {
                    is_fragment_input(declaration)
                        && declaration.is_statically_used
                        && !declaration.name.starts_with("gl_")
                        && !declaration.has_layout_qualifier
                });
                for fragment_input in fragment_inputs {
                    let mismatch = match vertex_outputs.get(fragment_input.name.as_str()) {
                        None => Some("no matching vertex output".to_string()),
                        Some(vertex_output)
                            if vertex_output.type_name != fragment_input.type_name =>
                        {
                            Some(format!(
                                "type {} in the vertex shader but {} in the fragment shader",
                                vertex_output.type_name, fragment_input.type_name
                            ))
                        }
                        Some(vertex_output)
                            if has_array_dimension_mismatch(
                                vertex_output.array_size,
                                fragment_input.array_size,
                            ) =>
                        {
                            Some("different array dimensions".to_string())
                        }
                        Some(vertex_output)
                            if vertex_output.interpolation != fragment_input.interpolation =>
                        {
                            Some(format!(
                                "interpolation {} in the vertex shader but {} in the fragment shader",
                                vertex_output.interpolation, fragment_input.interpolation
                            ))
                        }
                        Some(_) => None,
                    };
                    let Some(mismatch) = mismatch else {
                        continue;
                    };
                    let utf16_offset = fragment_shader.text[..fragment_input.byte_offset]
                        .encode_utf16()
                        .count();
                    ctx.diagnostic(
                    OxcDiagnostic::error(format!(
                        "Fragment input {} has {mismatch}, so the shader program cannot link with a defined value",
                        fragment_input.name
                    ))
                    .with_label(fragment_shader.origin_span(utf16_offset)),
                );
                }
            },
        );
    }
}

fn is_vertex_output(declaration: &GlslGlobalDeclaration) -> bool {
    declaration.qualifiers.contains("out") || declaration.qualifiers.contains("varying")
}

fn is_fragment_input(declaration: &GlslGlobalDeclaration) -> bool {
    declaration.qualifiers.contains("in") || declaration.qualifiers.contains("varying")
}
